### PR TITLE
Disable reference conversion during Solution Close operations

### PR DESCRIPTION
**Pull request by sharwell**
This change eliminates 96% (400MB&rarr;16MB) of the allocations and 94% of the time (4022ms&rarr;258ms) spent inside `CPSProject.Dispose` during solution close (including the IDE shutdown case).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Just to make sure: this will do the right thing when the user is just closing the solution (but not closing VS itself)?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -1815,21 +1823,24 @@ public void RemoveProjectOutputPath(ProjectId projectId, string outputPath)
                 projectReferenceInformation.OutputPaths.Remove(outputPath);
                 _projectsByOutputPath.MultiRemove(outputPath, projectId);
 
-                if (_projectsByOutputPath.TryGetValue(outputPath, out var remainingProjectsForOutputPath))
+                if (!_solutionClosing)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:1826]: doc.

**Comment by CyrusNajmabadi** [comment]
>  and 94% of the time (4022ms→258ms) spent inside CPSProject.Dispose during solution close 

I'm curious where costs are after this change.  Does Roslyn have other large hotspots we might want to look at next?

**Comment by davkean** [comment]
Nice! We probably have a bunch of this long hanging fruit across the product.

Just as an FYI, this will also fix VS branch switching cases where the solution or non-CPS based project has been touched; in those cases they reload the entire solution.

I'll also verified that we don't cancel close this after we've set it: https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fmsenv%2Fcore%2Fvsslnpst.cpp&version=GBmain&line=20080&lineEnd=20081&lineStartColumn=1&lineEndColumn=1&lineStyle=plain.

**Comment by sharwell** [comment]
> I'm curious where costs are after this change. Does Roslyn have other large hotspots we might want to look at next?

Roslyn contributed in two ways:

* CPU usage in reference conversion
* GC overhead throughout solution close, which primarily turned out to be allocations in reference conversion

For me, this change drops solution close from ~12 seconds to ~8 seconds. Significant work is still done by custom COM handling in the running document table. If Roslyn wanted to make further changes, we could suspend the Solution Crawler as soon as we get a Solution Closing notification.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:1834]: done!

**Comment by chsienki** [comment]
Nice! This always surprised me that it took so long on large solutions, makes sense why now, but good to see we're actively improving on it too :)

**Comment by sharwell** [comment]
> Just to make sure: this will do the right thing when the user is just closing the solution (but not closing VS itself)?

It _seems_ to be doing the right thing, but it wasn't a great experience before this change. I tried closing and reopening Roslyn repeatedly and occasionally it would have evaluation failures but nothing unique to this change.

**Comment by panopticoncentral** [comment]
Thanks for the fast response! I added a tracking bug so we can track in our backlog.

**Comment by CyrusNajmabadi** [comment]
> we could suspend the Solution Crawler as soon as we get a Solution Closing notification.

@mavasani  for awareness.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:1834]: perfect. t hanks :)

**Comment by jasonmalinowski** [comment]
@sharwell Do you know what the slow parts _were_ of the conversion? There's other cases where we can't optimize this in the same way and I wonder if we can address those directly too.

**Comment by davkean** [comment]
Filed a bug to represent that suggestion: https://github.com/dotnet/roslyn/issues/47408.

**Comment by sharwell** [comment]
@jasonmalinowski it's actually quite fast since we optimized dependency graph updates. The slowness was really an aggregate issue. If there's a specific scenario you'd like to measure we can look for targeted improvements.

**Pull request merged**